File: 15-3sum/3sum.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        nums.sort()
        result = []
        n = len(nums)

        for i in range(n-2):
            if i > 0 and nums[i] == nums[i-1]:
                continue

            j = i + 1
            k = n - 1

            while j < k:
                target = nums[i] + nums[j] + nums[k]
                if target == 0:
                    result.append([nums[i], nums[j], nums[k]])

                    while j < k and nums[j] == nums[j+1]:
                        j += 1
                    while j < k and nums[k] == nums[k-1]:
                        k -= 1

                    j += 1
                    k -= 1

                if target < 0:
                    while j < k and nums[j] == nums[j+1]:
                        j += 1
                    j += 1

                if target > 0:
                    while j < k and nums[k] == nums[k-1]:
                        k -= 1
                    k -= 1
        return result

File: 15-3sum/3sum.cpp
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        vector<vector<int>> res;
        int n = nums.size(), l, r, s;
        sort(nums.begin(), nums.end());

        for(int i = 0; i < n; i++) {
            if(nums[i] > 0) break;
            if(i > 0 && nums[i] == nums[i-1]) continue;
            
            l = i + 1;
            r = n - 1;
            while(l < r) {
                s = nums[i] + nums[l] + nums[r];
                if(s == 0) {
                    res.push_back({nums[i], nums[l], nums[r]});
                    l++;
                    r--;
                    while(l < r && nums[l] == nums[l-1]) l++;
                    while(l < r && nums[r] == nums[r+1]) r--;
                }
                else if(s > 0) {
                    r--;
                }
                else {
                    l++;
                }
            }

        }
        return res;
    }
};